iati/iati/custom_middleware.py

```diff
@@ -0,0 +1,20 @@
+"""Middleware for redirecting uppercase urls into lowercase."""
+from django import http
+from django.conf import settings
+
+
+class LowercaseMiddleware:
+    """Middleware class."""
+
+    def __init__(self, get_response):
+        """Initialise."""
+        self.get_response = get_response
+
+    def __call__(self, request):
+        """Redirect url paths as lowercase except for documents or media files."""
+        response = self.get_response(request)
+        path = request.get_full_path()
+        lower = path.lower()
+        if lower != path and not("/documents/" in path or settings.MEDIA_URL in path):
```

Comment:
These are some cases in which these conditions may cause a false positive.

For example, it isn't an impossible thought that 'Documents' or 'Media' Wagtail pages could be published, say:
https://iatistandard.org/en/about/case-studies/documents/
https://iatistandard.org/en/about/press-enquries/media/

These would therefore be excluded, and not redirected.

So, it might be a good idea to validate the 'excluded' URLs further. Also worth thinking about how these multiple conditionals can be readable.

Reply:
Furthermore, seeing as though this is bespoke behaviour, I'd suggest writing some tests.